File: C++Algorithm/DFS/6603/6603.cpp
#include <iostream>

using namespace std;

int arr[49];
int result[6];
int k = -1;

void dfs(int s, int depth)
{
    if (depth == 6) {
        for (int i = 0; i < 6; i++) {
            cout << result[i] << " ";
        }
        cout << endl;
        return;
    }

    for (int i = s; i < k; i++) {
        result[depth] = arr[i];
        dfs(i + 1, depth + 1);
    }
}

int main(int argc, char** argv)
{

    while (1) {
        cin >> k;
        if (k == 0)
            break;
        for (int i = 0; i < k; i++) {
            cin >> arr[i];
        }

        dfs(0, 0);
        cout << endl;
    }

    return 0;
}

File: Algorithm_python/Baekjun/Category/Backtracking/6603.py
#
# 7 1 2 3 4 5 6 7
# 8 1 2 3 5 8 13 21 34
# 0

def dfs(arr, result, idx, depth):
    if len(result) == 6:
        # print(result)
        print(' '.join(map(str, result)))
        return
    for i in range(idx, depth):
        result.append(arr[i])
        dfs(arr, result, i+1, depth)
        result.pop()


while True:
    arr = list(map(int, input().split()))
    if arr[0] == 0:
        break
    k = arr[0]
    result = []
    dfs(arr[1:],result, 0, k)
    print()
